Add a tooltip delay setting for editor

Fermats-Fish opened the pull request:
Resolves #27879 and resolves #9030.

Adds an editor setting which changes the tooltip delay for tooltips inside the editor.

A restart is required for this change to be applied, however the code could be extended so that this is not required.

This is my first time contributing to a major open source project, and my first time using C++, so let me know if I should have done anything differently.

Fermats-Fish committed:
Added tooltip delay setting for editor

Resolves #27879 and #9030.

Adds an editor setting which changes the tooltip delay for tooltips
inside the editor.

A restart is required for this change to be applied, however the
code could be extended so that this is not required.

YeldhamDev commented:
Would be nice to have this as an exposed property in `Viewport` for developers.

Fermats-Fish commented:
@YeldhamDev Do you mean using `ADD_PROPERTY` instead of `gui.tooltip_delay` so that game developers can change the tooltip delay for each viewport object?

YeldhamDev commented:
@Fermats-Fish Nevermind, completely forgot about the project setting.

Fermats-Fish commented:
@YeldhamDev No problem.

groud commented:
Honestly I am not sure this is worth it. It's yet another almost pointless settings...
Making it configurable for games makes sense though.

Fermats-Fish commented:
@groud At the very least it makes sense to make it so that changing the in game tooltip delay via project settings does not change the tooltip delay in the editor. I could always modify this code to just fix that issue without adding the editor setting.

groud commented:
Sure ! My single point was about the editor settings, but the rest of the change look fine to me :)

Calinou commented on scene/main/viewport.cpp:
```diff
@@ -3315,6 +3315,9 @@ void Viewport::_subwindow_visibility_changed() {
 	gui.subwindow_visibility_dirty = true;
 }
 
+float Viewport::editor_tooltip_delay = 0.5f;
+float Viewport::regular_tooltip_delay = 0.5f;
```
Since the `master` branch now accepts C++17 code, you can initialize those variables directly in the header file.

Also, is `static` needed here? We only make variables `static` if we absolutely need it.

Calinou commented on scene/main/viewport.h:
```diff
@@ -305,7 +308,7 @@ class Viewport : public Node {
 		Control *drag_preview;
 		float tooltip_timer;
-		float tooltip_delay;
+		float *tooltip_delay;
```
Out of curiosity, why was this `*` added?

Mickeon commented:
I was looking for this setting and I was astounded I could not find it. This may be salvaged, probably...

rsubtil commented:
+1 for salvaging; we want to have near instant (<0.05s) tooltips in our project, but we found out this property propagates to the editor, and with very low times it makes mouse scroll extremely finicky and annoying.

Is there any suggested workflow to salvage PRs? I assume I'd need to open another PR, would marking the current PR owner as `Co-authored-by:` be a good practice?

Calinou commented:
> Is there any suggested workflow to salvage PRs? I assume I'd need to open another PR, would marking the current PR owner as Co-authored-by: be a good practice?

Sure :slightly_smiling_face:

Fermats-Fish commented:
Hi everyone. If the community is interested in getting this feature merged now, I'm happy to have a go at fixing up my commit :).

Fermats-Fish commented:
Finally got some time to have a go at this.

I've fixed the merge conflicts, changed to the new system for creating editor settings, fixed the unnecessary pointer, and simplified the code by removing an unnecessary static variable.

The code still requires a restart every time you change the editor tooltip delay however. When I next get some time I can try to figure out a way to resolve this.

Calinou reviewed (changes requested):
Tested locally (rebased on top of `master` 9050ee1542f4e071188e8e4f868f3507bb31b3dc), there's an issue with how `gui/timers/tooltip_delay_sec` is registered. Also, the editor setting doesn't have any affect even after restarting the editor (I tried a tooltip delay of `0.05` and tooltips still take half a second to appear).

Since `GLOBAL_DEF()` is only reached when not running in the editor, this means the project setting is never registered in the editor, so you can't edit its value. It also won't appear in the generated class reference.

The `GLOBAL_DEF()` line should be moved to this location: https://github.com/godotengine/godot/blob/9050ee1542f4e071188e8e4f868f3507bb31b3dc/core/config/project_settings.cpp#L1468

After doing this, replace the call to `GLOBAL_DEF()` in `viewport.cpp` with `GLOBAL_GET()`.

akien-mga commented:
Superseded by #85678. Thanks for the contribution!
